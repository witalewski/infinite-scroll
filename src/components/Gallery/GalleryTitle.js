import React from 'react';

export const GalleryTitle = ({ title }) => (
  <div className="row">
    <div className="col">
      <h1>{title}</h1>
    </div>
  </div>
);
