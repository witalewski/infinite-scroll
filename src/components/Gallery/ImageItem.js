import React, { Component } from 'react';
import { observer, inject } from 'mobx-react';
import styled from '@emotion/styled';

const ImageItemStyled = styled.div`
  position: relative;
  width: 100%;
  height: 100%;

  .image {
    opacity: 0;
    position: absolute;
  }

  .image--loaded {
    opacity: 1;
    transition: opacity 0.5s;
  }

  .overlay {
    position: absolute;
    top: 50%;
    width: 100%;
    height: 40%;

    display: flex;
    justify-content: center;
    align-items: center;

    opacity: 1;

    background: rgba(255, 255, 255, 0.9);

    transition: opacity 0.2s;
    transform: translateY(-50%);
  }

  .overlay--hidden {
    opacity: 0;
  }

  .primary-button--hidden {
    display: none;
  }

  .message {
    width: 100%;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
  }

  .message--hidden {
    display: none;
  }

  .favourites-indicator {
    position: absolute;
    bottom: 12px;
    right: 12px;
    transition: opacity 0.5s 0.2s;
  }

  .favourites-indicator--hidden {
    opacity: 0;
  }
`;

export class ImageItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isImageLoaded: false,
      mouseOver: false,
      message: '',
      primaryButtonVisible: false,
      imageHeight: 0,
    };
  }

  componentWillUnmount() {
      if (this.timer) {
          clearTimeout(this.timer);
      }
  }
  onImageLoad = ({ target }) => {
    this.setState({ isImageLoaded: true , imageHeight: target.clientHeight});
  };

  onMouseOver = () => {
    this.setState({ mouseOver: true });
  };

  onMouseOut = () => {
    if (!this.state.message) {
      this.setState({
        mouseOver: false,
        message: '',
        undoVisible: false,
        undoAction: null,
      });
    }
  };

  addToFavourites = () => {
    this.props.addToFavourites(this.props.image);
    this.setState({
      message: 'Added to favourites!',
      undoAction: this.removeFromFavourites,
    });
    if (this.timer) {
      clearTimeout(this.timer);
    }
    this.timer = setTimeout(() => {
      this.setState({
        mouseOver: false,
        message: '',
        undoVisible: false,
        undoAction: null,
        primaryButtonVisible: false,
      });
    }, 2000);
  };

  removeFromFavourites = () => {
    this.props.removeFromFavourites(this.props.image);
    this.setState({
      message: 'Removed from favourites.',
      undoAction: this.addToFavourites,
    });
    if (this.timer) {
      clearTimeout(this.timer);
    }
    this.timer = setTimeout(() => {
      this.setState({
        mouseOver: false,
        message: '',
        undoVisible: false,
        undoAction: null,
        primaryButtonVisible: false,
      });
    }, 2000);
  };

  render() {
    const { image, favourites } = this.props;
    const { mouseOver, isImageLoaded, imageHeight } = this.state;
    const isFavourite = favourites.includes(this.props.image);
    return (
      <ImageItemStyled
        onMouseOver={this.onMouseOver}
        onMouseOut={this.onMouseOut}
        style={{height: imageHeight || this.props.height}}
      >
        {image && (
          <img
            className={`image ${isImageLoaded && 'image--loaded'}`}
            alt="Shibe dog"
            src={image}
            decoding="async"
            onLoad={this.onImageLoad}
          />
        )}
        <div className={`overlay ${!mouseOver && 'overlay--hidden'}`}>
          <button
            type="button"
            className={`btn btn-outline-primary primary-button ${this.state
              .message && 'primary-button--hidden'}`}
            onClick={
              isFavourite ? this.removeFromFavourites : this.addToFavourites
            }
          >
            {isFavourite ? 'üíî Remove from Favourites' : '‚ù§Ô∏è Add to Favourites'}
          </button>
          <div
            className={`message ${!this.state.message && 'message--hidden'}`}
          >
            <span>{this.state.message}</span>
            <button
              type="button"
              className="btn btn-secondary btn-sm"
              onClick={this.state.undoAction}
            >
              Undo
            </button>
          </div>
        </div>
        <div
          className={`favourites-indicator ${!isFavourite &&
            'favourites-indicator--hidden'}`}
        >
          <span role="img" aria-label="favourite">
            ‚ù§Ô∏è
          </span>
          Ô∏è
        </div>
      </ImageItemStyled>
    );
  }
}

export default inject(({ appState }) => ({
  favourites: appState.favourites,
  addToFavourites: appState.addToFavourites,
  removeFromFavourites: appState.removeFromFavourites,
}))(observer(ImageItem));
