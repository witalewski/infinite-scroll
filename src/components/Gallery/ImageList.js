import React, { Component } from 'react';
import styled from '@emotion/styled';
import ImageItem from './ImageItem/ImageItem';
import { placeholderHeightGenerator } from '../../utils/placeholderHeightGenerator';
import { LIST_MARGIN } from '../../global/constants';

const ImageListStyled = styled.ul`
    list-style: none;
    display: flex;
    flex-direction: column;
    flex-basis: 50%;
    padding: 0;
    margin: 0 ${LIST_MARGIN}px;
`;
export class ImageList extends Component {
  render() {
    const { list, columnWidth } = this.props;
    return (
      <ImageListStyled className="image-list">
        {list.map((image, i) => (
          <li
            className="image-list--item"
            key={`${image || 'placeholder'}-${i}`}
          >
            <ImageItem
              image={image}
              height={placeholderHeightGenerator.next() * columnWidth}
            />
          </li>
        ))}
      </ImageListStyled>
    );
  }
}
