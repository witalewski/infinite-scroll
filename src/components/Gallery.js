import React, { Component, createRef } from 'react';
import { List, Range } from 'immutable';
import styled from '@emotion/styled';
import { ImageList } from './ImageList';
import {
  LIST_MARGIN,
  MIN_COLUMN_WIDTH,
} from '../global/constants';

const PhotosViewStyled = styled.div`
  .image-list-container {
    display: flex;
    justify-content: center;
    align-items: flex-start;
  }

  .image-list {
    list-style: none;
    display: flex;
    flex-direction: column;
    flex-basis: 50%;
    padding: 0;
    margin: 0 ${LIST_MARGIN}px;
  }

  .image {
    width: 100%;
    object-fit: cover;
  }

  .image-placeholder {
    width: 100%;
    height: 100%;
  }
`;

export class Gallery extends Component {
  constructor(props) {
    super(props);
    this.listContainerRef = createRef();
    this.state = {
      initialized: false,
      columnsCount: 0,
      actualColumnWidth: 0,
    };
  }

  getMinColumnHeight = () => {
    const columns = this.listContainerRef.current.childNodes;
    if (columns && columns.length > 0) {
      let minHeight = Infinity;
      for (let i = 0; i < columns.length; i++) {
        minHeight = Math.min(minHeight, columns[i].clientHeight);
      }
      return minHeight;
    }
    return 0;
  };

  onScroll = () => {
    const scrollPosition =
      window.pageYOffset || document.documentElement.scrollTop;
    if (this.getMinColumnHeight() < window.innerHeight + scrollPosition) {
      this.props.requestMoreImages(16);
    }
  };

  onResize = () => {
    const columnsCount = Math.floor(
      this.listContainerRef.current.clientWidth / MIN_COLUMN_WIDTH
    );
    const actualColumnWidth =
      this.listContainerRef.current.clientWidth / columnsCount -
      2 * LIST_MARGIN;
    this.setState({
      columnsCount,
      actualColumnWidth,
    });
  };

  componentDidMount() {
    if (!this.state.initialized) {
      window.addEventListener('scroll', this.onScroll);
      window.addEventListener('resize', this.onResize);

      //invoke at once to get initial values
      this.onScroll();
      this.onResize();

      this.setState({ initialized: true });
    }
  }

  render() {
    const { columnsCount, actualColumnWidth } = this.state;
    const { images } = this.props;
    const imagesByColumn = List(Range(0, columnsCount).map(_ => []));
    if (columnsCount) {
      images.forEach((elem, i) =>
        imagesByColumn.get(i % columnsCount).push(elem)
      );
    }
    return (
      <PhotosViewStyled className="row">
        <div className="col image-list-container" ref={this.listContainerRef}>
          {imagesByColumn.map((list, i) => (
            <ImageList
              key={`list-${i}-of-${columnsCount}`}
              list={list}
              columnWidth={actualColumnWidth}
            />
          ))}
        </div>
      </PhotosViewStyled>
    );
  }
}
