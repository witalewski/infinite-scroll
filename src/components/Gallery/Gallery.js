import React, { Component, createRef } from 'react';
import { List, Range } from 'immutable';
import styled from '@emotion/styled';
import { ImageList } from './ImageList';
import { GalleryTitle } from './GalleryTitle';
import { LIST_MARGIN, MIN_COLUMN_WIDTH } from '../../global/constants';

const PhotosViewStyled = styled.div`
  .image-list-container {
    display: flex;
    justify-content: center;
    align-items: flex-start;
  }
`;

export class Gallery extends Component {
  constructor(props) {
    super(props);
    this.listContainerRef = createRef();
    this.state = {
      initialized: false,
      columnsCount: 0,
      actualColumnWidth: 0,
    };
  }

  getMinColumnHeight = () => {
    const columns = this.listContainerRef.current.childNodes;
    let minHeight = Infinity;
    if (columns && columns.length > 0) {
      for (let i = 0; i < columns.length; i++) {
        minHeight = Math.min(minHeight, columns[i].clientHeight);
      }
    } else {
      minHeight = 0;
    }
    return minHeight;
  };

  onScroll = () => {
    const scroll = window.pageYOffset || document.documentElement.scrollTop;
    this.getMinColumnHeight() < window.innerHeight + scroll &&
      this.props.requestMoreImages();
  };

  onResize = () => {
    const containerWidth = this.listContainerRef.current.clientWidth;
    const columnsCount = Math.max(
      Math.floor(containerWidth / MIN_COLUMN_WIDTH),
      1
    );
    const columnWidth = containerWidth / columnsCount - 2 * LIST_MARGIN;
    this.setState({
      columnsCount,
      columnWidth,
    });
  };

  setupResizeHandlers() {
    window.addEventListener('resize', this.onResize);
    this.onResize();
  }

  setupScrollHandlers() {
    window.addEventListener('scroll', this.onScroll);
    this.onScroll();
  }

  componentDidMount() {
    if (!this.state.initialized) {
      this.setupResizeHandlers();
      this.props.infinite && this.setupScrollHandlers();
      this.setState({ initialized: true });
    }
  }
  componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
    this.props.infinite && window.removeEventListener('scroll', this.onScroll);
  }

  render() {
    const { columnsCount, columnWidth } = this.state;
    const { images, title } = this.props;
    const imagesByColumn = List(Range(0, columnsCount).map(_ => []));

    if (columnsCount && columnWidth) {
      images.forEach((elem, i) => {
        imagesByColumn.get(i % columnsCount).push(elem);
      });
    }

    return (
      <PhotosViewStyled>
        <GalleryTitle title={title} />
        <div className="row">
          <div className="col image-list-container" ref={this.listContainerRef}>
            {imagesByColumn.map((list, i) => (
              <ImageList
                key={`list-${i}-of-${columnsCount}`}
                list={list}
                columnWidth={columnWidth}
              />
            ))}
          </div>
        </div>
      </PhotosViewStyled>
    );
  }
}
