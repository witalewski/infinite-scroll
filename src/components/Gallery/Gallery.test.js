import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import mockImageURLs from '../../fixtures/imageURLs.json';
import { Gallery } from './Gallery';

describe('Gallery', () => {
  it('matches snapshot', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <Gallery
        infinite
        title="Test Gallery"
        images={mockImageURLs}
        requestMoreImages={jest.fn()}
      />
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
