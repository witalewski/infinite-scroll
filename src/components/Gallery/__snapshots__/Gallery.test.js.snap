// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery matches snapshot 1`] = `
<ForwardRef(render)>
  <GalleryTitle
    title="Test Gallery"
  />
  <div
    className="row"
  >
    <div
      className="col image-list-container"
    >
      Immutable.List []
    </div>
  </div>
</ForwardRef(render)>
`;
